CONVENTIONS.md

```diff
@@ -85,3 +85,86 @@ metal3-io](https://github.com/metal3-io/metal3-docs/blob/master/design/bare-meta
 ### API
 
 OpenShift APIs follow the [Kubernetes API conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md).
+
+### Operators
+
+The OpenShift project is an early adopter of, and makes extensive use of, [the
+operator pattern](https://kubernetes.io/docs/concepts/extend-kubernetes/operator/),
+and so it is incumbent on us to establish some conventions around operators.
+
+#### Taints and Tolerations
+
+An operator deployed by the CVO should run on master nodes and therefore should
+tolerate the following taint:
+
+* `node-role.kubernetes.io/master`
+
+For example:
+
+```yaml
+spec:
+  template:
+    spec:
+      tolerations:
+      - key: node-role.kubernetes.io/master
+        operator: Exists
+        effect: NoSchedule
+```
+
+Tolerating this taint should suffice for the vast majority of core OpenShift
+operators.  In exceptional cases, an operator may tolerate one or more of the
+following taints if doing so is necessary to form a functional Kubernetes node:
+
+* `node.kubernetes.io/disk-pressure`
+* `node.kubernetes.io/memory-pressure`
+* `node.kubernetes.io/network-unavailable`
+* `node.kubernetes.io/not-ready`
+* `node.kubernetes.io/pid-pressure`
+* `node.kubernetes.io/unreachable`
+
+Operators should not specify tolerations in their manifests for any of the taints in
+the above list without an explicit and credible justification.
+
+When an operator configures its operand, the operator likewise may specify
+tolerations for the aforementioned taints but should do so only as necessary and only
+with explicit justification.
+
+Note that the DefaultTolerationSeconds and PodTolerationRestriction admission plugins
+may add time-bound tolerations to an operator or operand in addition to the
+tolerations that the operator has specified.
+
+If appropriate, a CRD that corresponds to an operand may provide an API to allow
+users to specify a custom list of tolerations for that operand.  For examples, see
+the
+[imagepruners.imageregistry.operator.openshift.io/v1](https://github.com/openshift/api/blob/34f54f12813aaed8822bb5bc56e97cbbfa92171d/imageregistry/v1/types_imagepruner.go#L67-L69),
+[configs.imageregistry.operator.openshift.io/v1](https://github.com/openshift/api/blob/34f54f12813aaed8822bb5bc56e97cbbfa92171d/imageregistry/v1/types.go#L82-L84),
+[builds.config.openshift.io/v1](https://github.com/openshift/api/blob/34f54f12813aaed8822bb5bc56e97cbbfa92171d/config/v1/types_build.go#L96-L99),
+and
+[ingresscontrollers.operator.openshift.io/v1](https://github.com/openshift/api/blob/34f54f12813aaed8822bb5bc56e97cbbfa92171d/operator/v1/types_ingress.go#L183-L191)
+APIs.
+
+In exceptional cases, an operand may tolerate all taints:
+
+* if the operand is required to form a functional Kubernetes node, or
```

Comment:
Apologies for being late on reviewing this PR.

To me this doesn't warrant operand to have blanket tolerations. This means we should be explicit on the tolerations that need to matched for the pod to keep running. 

Some of the side effects that we noticed with this blanket toleration are
- node controller wouldn't evict the pods when the node is marked as not ready. We faced this in the [past](https://bugzilla.redhat.com/show_bug.cgi?id=1705649). So allowing allowing blanket tolerations might cause problems in future.
- In future, we'd like to add Windows nodes to OpenShift clusters. The current way to ensure that the Windows workloads lands on Windows nodes and linux workloads lands on linux nodes is via nodeSelector and taints&tolerations. In a way, with this we are allowing operand or operator to tolerate Windows Specific taints and let them land on the Windows nodes which will cause problems.

Reply:
cc @aravindhp @gmarkley-VI

Reply:
I agree with @ravisantoshgudimetla. Having an operand tolerate all taints will definitely cause issues on clusters that have enabled Windows workloads.

Reply:
> To me this doesn't warrant operand to have blanket tolerations. This means we should be explicit on the tolerations that need to matched for the pod to keep running.

Users can set arbitrary taints, so prohibiting operands from specifying blanket tolerations would require either (a) requiring users to specify operands' tolerations too or (b) restricting the user to some set of predefined taints.

> Some of the side effects that we noticed with this blanket toleration are
> 
> * node controller wouldn't evict the pods when the node is marked as not ready. We faced this in the [past](https://bugzilla.redhat.com/show_bug.cgi?id=1705649). So allowing allowing blanket tolerations might cause problems in future.

That Bugzilla report seems to be about operators, not about operands that need to run on each node.  For example, cluster-dns-operator only needs to run on some node somewhere, but the dns-node-resolver container in its daemonset operand must run any given node in order for that node to be fully functional; does specifying a blanket toleration on that operand cause a problem?

> * In future, we'd like to add Windows nodes to OpenShift clusters. The current way to ensure that the Windows workloads lands on Windows nodes and linux workloads lands on linux nodes is via nodeSelector and taints&tolerations. In a way, with this we are allowing operand or operator to tolerate Windows Specific taints and let them land on the Windows nodes which will cause problems.

Why do you need both a node selector and a taint?  Why isn't a node selector sufficient?  For example, cluster-dns-operator creates a daemonset that tolerates all taints but also specifies a node selector: `kubernetes.io/os: linux`.  Would it help to add a section on node selectors with a statement that an operator or operand that requires a specific operating system must specify an appropriate node selector?